OpenRA.Mods.Common/DiscordService.cs

```diff
@@ -0,0 +1,197 @@
+			client.GetType().GetProperty("HasRegisteredUriScheme").SetValue(client, true);
+
+			client.SetSubscription(EventType.Join | EventType.JoinRequest);
+			client.Initialize();
+		}
+
+		void OnJoinRequested(object sender, JoinRequestMessage args)
+		{
+			var client = (DiscordRpcClient)sender;
+			client.Respond(args, true);
+		}
+
+		void OnJoin(object sender, JoinMessage args)
+		{
+			if (currentState != DiscordState.InMenu)
+				return;
+
+			var server = args.Secret.Split('|');
+			Game.RunAfterTick(() =>
+			{
+				Game.RemoteDirectConnect(new ConnectionTarget(server[0], int.Parse(server[1])));
+			});
+		}
+
+		void SetStatus(DiscordState state, string details = null, string secret = null)
+		{
+			if (currentState == state)
+				return;
+
+			if (Service.Value == null)
+				return;
+
+			string stateText;
+			DateTime? timestamp = null;
+			Party party = null;
+			Secrets secrets = null;
+
+			switch (state)
+			{
+				case DiscordState.InMenu:
+					stateText = "In menu";
+					break;
+				case DiscordState.InMapEditor:
+					stateText = "In Map Editor";
+					break;
+				case DiscordState.InSkirmishLobby:
+					stateText = "In Skirmish Lobby";
+					break;
+				case DiscordState.InMultiplayerLobby:
+					stateText = "In Multiplayer Lobby";
+					timestamp = DateTime.UtcNow;
+					party = new Party
+					{
+						ID = Secrets.CreateFriendlySecret(new Random()),
+						Size = 1,
+						Max = 1
+					};
+					secrets = new Secrets
+					{
+						JoinSecret = secret
+					};
+					break;
+				case DiscordState.PlayingMultiplayer:
+					stateText = "Playing Multiplayer";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.PlayingCampaign:
+					stateText = "Playing Campaign";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.WatchingReplay:
+					stateText = "Watching Replay";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.PlayingSkirmish:
+					stateText = "Playing Skirmish";
+					timestamp = DateTime.UtcNow;
+					break;
+				default:
+					throw new ArgumentOutOfRangeException("state", state, null);
+			}
+
+			var richPresence = new RichPresence
+			{
+				Details = details,
+				State = stateText,
+				Assets = new Assets
+				{
+					LargeImageKey = "large",
+					LargeImageText = Game.ModData.Manifest.Metadata.Title,
+				},
+				Timestamps = timestamp.HasValue ? new Timestamps(timestamp.Value) : null,
+				Party = party,
+				Secrets = secrets
+			};
+
+			client.SetPresence(richPresence);
+			currentState = state;
+		}
+
+		public void Dispose()
+		{
+			client.Dispose();
+		}
+
+		public static void UpdateStatus(DiscordState state, string details = null, string secret = null)
+		{
+			if (Service.Value != null)
+				Service.Value.SetStatus(state, details, secret);
+		}
+
+		public static void SetPlayers(int players, int slots)
+		{
+			if (Service.Value != null)
```

Comment:
Style nit: this is semantically a single statement, but should still have braces to avoid confusion with the hanging block.